sympy/polys/polytools.py

```diff
@@ -5497,7 +5497,11 @@ def terms_gcd(f, *gens, **args):
     from sympy.core.relational import Equality
 
     orig = sympify(f)
-    if not isinstance(f, Expr) or f.is_Atom:
+
+    # XXX: Should any Relational be accepted here? Why does this accept
+    # any Relational at all? Perhaps it should not and any code that calls
+    # this with Relational should be changed.
+    if not isinstance(f, (Expr, Relational)) or f.is_Atom:
```

Comment:
I think that Equality is the only Relational that could be considered as a polynomial expression (namely lhs - rhs) and even that is returned intact later. Maybe all Relationals should be returned here.

Reply:
I'm not sure I understand the intent of this. Would it not be better to raise when given a non-Expr?

Reply:
Neither do I. I think that it could raise for a non-Expr other than Equality that is often treated as an expression.

Reply:
I've made this raise on inequalities. Also I changed it to just do:
```
if instance(Eq):
    return Eq(terms_gcd(lhs), terms_gcd(rhs))
```
since that is what this is being used for (in constantsimp for ODEs). The only other place to call this with an Eq a couple of times in the terms_gcd tests.

Reply:
Another way that this goes wrong on master is
```julia
In [1]: Integral(Eq(x, y))                                                                                                        
Out[1]: 
⌠        ⌠     
⎮ x dx = ⎮ y dy
⌡        ⌡ 
```
which has a different integration variable on each side of the equation.

I've added a deprecation warning for `Integral(Eq(x, y))`. I've also added a deprecation warning `Eq(x, y).integrate()` or `Eq(x, y).integrate(z)` without limits.

Reply:
There is also `Product(Eq(x, y))` and `Sum(Eq(x, y))`:
https://github.com/sympy/sympy/blob/1923822ddf8265199dbd9ef9ce09641d3fd042b9/sympy/concrete/tests/test_sums_products.py#L917-L921
The change here currently makes those give deprecation warning. There isn't an alternative method that could be used for those. Those are at least valid since they have limits.

Maybe it's better just to deprecate `Integral(Eq)` without limits. Then `.integrate` can still just call `integrate`.